workflow-format.md

```diff
@@ -0,0 +1,236 @@
+# OCR-D Workflow format
+> Sequence of [processor](glossary#processor) [calls](cli)
+
+Revision: 1
+
+## Rationale
+
+OCR consists of many [steps](glossary#activities) that need to be run in
+sequence ([*workflow*](glossary#workflow)). In OCR-D, these steps are
+implemented in [*procesors*](https://ocr-d.de/en/workflows), executables that
+read and write files to one or more `mets:fileGrp` in a METS document. While it
+is possible to define such a workflow as a sequence of commands in a shell
+script, that approach has many drawbacks (no checks for correct input/output
+wiring, no syntax checks, easy to end up with an invalid
+[workspace](glossary#workspace)).
+
+A subset of shell script is however a good solution to define a common,
+workflow-engine-agnostic syntax for workflows in OCR-D. It is straightforward
+to tokenize [according to
+POSIX](https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_03)
+(`sh`) and every processor invocation can be parsed according to the [OCR-D CLI
+spec](cli). Every step can be reproduced easily by executing it in a command
+shell.
+
+While such an OCR-D workflow description "script" (*OCRD-WF*) is syntactically
+compatible with a POSIX-compliant command shell, it has different semantics and
+supports only a small subset of shell script syntax and functionality.
+
+## File structure and syntax
+
+A OCRD-WF script has two logical sections: preamble and steps.
+
+The preamble contains the shebang and any variable
+assignments while the steps contain the processor calls. Both preamble and steps can
+contain any number of comment lines.
+
+### Shebang
+
+The first line of every OCRD-WF must match the regular expression
+`#!/usr/bin/env ocrd-wf-v[1-9]+[0-9]*\n` 
+
+The number represents the revision of this spec, so initially, all OCRD-WF
+should begin with `#!/usr/bin/env ocrd-wf-v1`.
+
+### Comments
+
+Everyline that begins with any whitespace followed by the character "#" (*pound
+sign*, `\x23`).
+
+### Line continuation
+
+If a line ends with "\" (*backslash*, `\x5c`), backslash and newline are
+removed and the next line, without leading whitespace, is appended. This is
+repeated while the joined ends with a backslash.
+
+### Variable Assignments
+
+An OCRD-WF can contain variable assignments of the form `name=value\n`. `name` must be
+a valid `sh` identifier, everything `=` is parsed as a string.
+
+<!-- TODO define behavior? -->
+**NOTE** The semantics of assigning `value` to `name` are implementation-dependant
+
+## Parsing Algorithm
+
+The parsing algorithm consists of a set of passes. It should inform
+implementers but does not require require an exact implementation, as long as
+the result is the same.
+
+**NOTE** Unlike in POSIX shell, comments are removed before line continuations are
+resolved, so you can intersperse line continuation with comments:
+
+```sh
+ocrd-foo \
+  # This parameter does bar
+  -P foo bar
+```
```

Comment:
It's a pet peeve of mine TBH that you cannot comment on continued lines in shell script. I understand that this is not POSIXly correct but how does it make implementation harder? If the first step in an implementation consists of removing empty lines and comments - and we allow only full-line comments - I think it makes implementation actually simpler.

I agree that it's probably a bad design decision to allow syntax that will trip up `/bin/sh` if you try to run an OCRD-WF as a shell script. But should you do that?

Reply:
> but how does it make implementation harder? If the first step in an implementation consists of removing empty lines and comments - and we allow only full-line comments - I think it makes implementation actually simpler.

It would interfere with the other provision about quoted newlines:
```sh
# a very long comment indeed which \
  needs to be continued (which is covered by POSIX)
echo first command
```
If you blindly strip away the first line, then the second becomes invalid. You could still try to do your own thing by first expanding backslash-newlines and then removing comments. But you cannot rely on standard implementations anymore.

(And I don't think it's a good idea to forbid continued comment lines, because its yet another nonstandard/unexpected syntax.)

> and is unusual/unexpected (not only to shell-experienced users but also editors that understand shell syntax and can highlight it).

Any comment on that aspect?

> I agree that it's probably a bad design decision to allow syntax that will trip up `/bin/sh` if you try to run an OCRD-WF as a shell script. But should you do that?

No, I wouldn't really care about that problem.

Reply:
> and is unusual/unexpected (not only to shell-experienced users but also editors that understand shell syntax and can highlight it).

Neither GitHub nor Vim correctly highlight comment continuations. POSIXly correct but not correctly implemented. Why would use those over just prepending `#` to the continued line?

```sh
# comment \
  continued
ls
```

However neither GH nor Vim has a problem with comments interspersed in comments.

```sh
ls \
  # comment
  continued
```

(not completely true, interspersed comments will mess with vim's autoindentation but not the syntax highlighting with `sh` filetype).

Reply:
> Neither GitHub nor Vim correctly highlight comment continuations. POSIXly correct but not correctly implemented.
> However neither GH nor Vim has a problem with comments interspersed in comments.

Huh, same for Emacs :-)

> Why would use those over just prepending `#` to the continued line?

I probably wouldn't, I was just arguing that deviating from standard sh could be surprising to users and implementors. IMHO comments on line continuations are not worth that deviation.

Reply:
> IMHO comments on line continuations are not worth that deviation.

I think they are. but maybe @cneud or @EEngl52 can be tiebreaker here - this detail should not prevent us from moving forward and I'm fine with scratching that feature if I'm alone in my stand against POSIX line continuation strictness :)

Reply:
I certainly wouldn't mind and actually for me it is easier to just use `#` for every line

Reply:
> I certainly wouldn't mind and actually for me it is easier to just use `#` for every line

My question was more about comments in continued lines, e.g

```
ocrd-foo \
  # documenting this specific parameter
  -P bla 42 \
  # and this input file group
  -I FOO
```

But I see no point of defending this if no one else sees the benefit. I'll drop that as an official "feature". As the parsing algorithm is defined at the moment, it would still work though, this needs to be aligned with POSIX then.